tests-fuzz/src/generator/create_expr.rs

```diff
@@ -200,6 +206,115 @@ impl<R: Rng + 'static> Generator<CreateTableExpr, R> for CreateTableExprGenerato
+#[derive(Builder)]
+#[builder(pattern = "owned")]
+pub struct CreateLogicalTableExprGenerator<R: Rng + 'static> {
+    table_ctx: TableContextRef,
+    labels: usize,
+    #[builder(default)]
+    _phantom: PhantomData<R>,
+}
+
+impl<R: Rng + 'static> Generator<CreateTableExpr, R> for CreateLogicalTableExprGenerator<R> {
+    type Error = Error;
+
+    fn generate(&self, rng: &mut R) -> Result<CreateTableExpr> {
+        // Currently we mock the usage of GreptimeDB as Prometheus' backend, the physical table must have two columns.
+        ensure!(
+            self.table_ctx.columns.len() == 2,
+            error::UnexpectedSnafu {
+                violated: "The physical table must have two columns"
+            }
+        );
+
+        // Generates the logical table columns based on the physical table.
+        let physical_table_name = self.table_ctx.name.to_string().replace('`', "");
+        let table_generator = CreateTableExprGeneratorBuilder::default()
+            .name_generator(Box::new(MappedGenerator::new(
+                WordGenerator,
+                merge_two_word_map_fn(random_capitalize_map, uppercase_and_keyword_backtick_map),
+            )))
+            .columns(self.labels)
+            .engine("metric")
+            .column_type_generator(Box::new(StringColumnTypeGenerator))
+            .column_options_generator(Box::new(partible_column_options_generator))
+            .with_clause([(
+                "on_physical_table".to_string(),
+                physical_table_name.to_string(),
+            )])
+            .build()
+            .unwrap();
+
+        let mut table = table_generator.generate(rng)?;
+        while table.table_name.value == physical_table_name {
+            table.table_name = table_generator.name_generator.gen(rng);
+        }
+        let logical_ts = table.columns.iter().position(|column| {
+            column
+                .options
+                .iter()
+                .any(|option| option == &ColumnOption::TimeIndex)
+        });
+        table.columns.remove(logical_ts.unwrap());
+        table.columns.iter_mut().for_each(|column| {
+            // Only keeps the primary key option for string columns.
+            column
+                .options
+                .retain(|option| option == &ColumnOption::PrimaryKey);
+            // Ensures the column name is unique.
+            while column.name.value == self.table_ctx.columns[0].name.value
+                || column.name.value == self.table_ctx.columns[1].name.value
+            {
+                column.name = table_generator.name_generator.gen(rng);
+            }
+        });
+
+        table.columns.extend(self.table_ctx.columns.clone());
+
+        let mut primary_keys = vec![];
+        for (idx, column) in table.columns.iter().enumerate() {
+            if column.is_primary_key() {
+                primary_keys.push(idx);
+            }
+        }
+        primary_keys.shuffle(rng);
+        table.primary_keys = primary_keys;
```

Comment:
Actually the primary key option is kept, the `not null` and default value would be removed. The tests in #3772 also add `primary key` option to the newly added string columns. If not, the column would be a field rather than a tag then refused by the validator in that patch. So maybe we need to keep this?

Reply:
Good catch. We are still the primary key option🥲.